### ISSUE TITLE
Can't use get_managed_token for system assigned MI or user assigned MI within App Service, error 404

### ISSUE THREAD

**Issue by dmuijen** [Azure/AzureAuth#75]
Hello,

Using a system or user assigned identity in combination with `get_managed_token` gives a 404 error. Could not get the function working from within an app service running a Shiny app using a managed identity to access a storage account. AzureAuth seems to make the wrong API call at and endpoint `/MSI/token/token` that does not exist and hence returns 404 since it should be `http://{endpoint}/MSI/token`. From within a VM for local development obtaining a token using managed identity using AzureAuth works as expected.

Current workaround for the app service is to create the API call manually instead of using AzureAuth for a system assigned MI reading data from a storage account in a Shiny app hosted in App Services.

```
library(httr)
library(AzureStor)

# Fetch / set variables for API call within app service running Shiny app running on Shiny server
api_version <- "2019-08-01" ## Might work with newer versions as well
mi_endpoint <- Sys.getenv("MSI_ENDPOINT")
mi_secret <- Sys.getenv("MSI_SECRET")
resource <- "https://storage.azure.com"
storage_account <- "https://<storageaccountname>.blob.core.windows.net"
container_name <- "mycontainer"
headers <- c(
`X-IDENTITY-HEADER` = mi_secret
)

## Fetch managed identity token
res <- GET(url = paste0(mi_endpoint, '?resource=', resource, "&api-version=", api_version), add_headers(.headers=headers))
mi_token <- content(res)$access_token

## Connect to storage account using MI
ad = storage_endpoint(storage_account, token = mi_token)
cont = storage_container(ad, container_name) 
```

**Comment by kmcoding** [comment]
I was experiencing the exact same issue as you described and your workaround is working flawlessly for us now! Thank you so much! And indeed the package is in dire need of an update...

**Comment by hongooi73** [comment]
The endpoint should already be taken from the MSI_ENDPOINT environment variable, the same as what you've got in your sample code. It gets the secret from the MSI_SECRET variable as well, but it puts it into the HTTP 'secret' header instead of 'X-IDENTITY-HEADER'. Are you sure you've got your variables set correctly? In particular, don't put the `token` bit at the end of MSI_ENDPOINT.

**Comment by dmuijen** [comment]
Below is untested but perhaps the second token part is added in this snippet in `AzureTokenManaged`. Reading the MSI endpoint sys variable with `Sys.getenv("MSI_ENDPOINT")` already gives an URL that ends in `/token` so adding another `/token` part to the uri might be the issue.

```
private=list(

    initfunc=function(init_args)
    {
        stopifnot(is.list(self$token_args))

        uri <- private$aad_uri("token") ### Here?
        query <- utils::modifyList(self$token_args,
            list(`api-version`=getOption("azure_imds_version"), resource=self$resource))

        secret <- Sys.getenv("MSI_SECRET")
        headers <- if(secret != "")
            httr::add_headers(secret=secret)
        else httr::add_headers(metadata="true")

        httr::GET(uri, headers, query=query)
    }
))
```

**Comment by hongooi73** [comment]
Yes, the `token` is added during processing. This works in my testing, but that was in an Azure VM, not an app service. So the MSI_ENDPOINT variable isn't under your control? Does it work with other code that uses the service?

**Comment by dmuijen** [comment]
Hello, ok thanks for the clarification. Within an App Service the system assigned MI MSI_ENDPOINT environment variable is set by azure and includes the `/token` bit in the URL already.

**Comment by drdrjacobs** [comment]
Are there any prospects for fixing this?

**Comment by lgaborini** [comment]
I'm encountering the same issue trying to authenticate with AzureKeyVault with a system-assigned MI!
The workaround must be slightly modified to work (running in an Azure Container Instance):

```r
api_version <- "2019-08-01"
mi_endpoint <- Sys.getenv("IDENTITY_ENDPOINT")
resource <- "https://my-keyvault.vault.azure.net"
headers <- c(
   `X-IDENTITY-HEADER` = Sys.getenv("IDENTITY_HEADER")
)

## Fetch managed identity token
res <- httr::GET(
   url = paste0(mi_endpoint, '?resource=', resource, "&api-version=", api_version),
   httr::add_headers(.headers = headers)
)
token <- httr::content(res)$access_token
```

[Reference](https://learn.microsoft.com/en-us/azure/app-service/overview-managed-identity?tabs=portal%2Chttp#connect-to-azure-services-in-app-code)

**Comment by hongooi73** [comment]
Sorry for the extended delay in responding. I've pushed a fix in the `managed-fix` branch. @lgaborini @drdrjacobs can you try it out and see if it works?

**Comment by lgaborini** [comment]
Working perfectly, thanks for the speedy fix!   
It's also picked up via `AzureKeyVault::key_vault()`.

The version in DESCRIPTION is not valid though:

```
> pak::pkg_install("Azure/AzureAuth@managed-fix")
Error:
! error in pak subprocess
Caused by error:
! invalid version specification ‘1.3.4-dev’
Type .Last.error to see the more details.
```

I replaced it with the conventional `1.3.4.9000`.

**Comment by hongooi73** [comment]
I suggesting telling the pak maintainers about that, actually. Semantic versioning only requires 3 numbers, the major, minor and and patch versions. These would be the "1.3.4" in the AzureAuth version. After that, any letters or digits should be allowed, eg you'll often see release candidates marked as "x.y.z-rc1" or something similar.
